### PR TITLE
feat(finetune): add DPO objective

### LINKED ISSUE #826: [REQ] add DPO fine-tuning objective
Parent: #214

Depends on #820 and #824.

## Summary

Add Direct Preference Optimization as a separate fine-tuning objective using canonical preference records. Dataset schema and training objective must remain separate concepts.

```yaml
apiVersion: v1alpha1
baseModel: unsloth/model

objective:
  type: dpo
  beta: 0.1
  lossType: sigmoid
  maxPromptLength: 512

datasets:
  - source: acme/preferences
    type: preference
    loader:
      type: huggingface
      split: train
      revision: 0123456789abcdef0123456789abcdef01234567
```

Initial record:

```json
{
  "prompt": "How should I rotate an API key?",
  "chosen": "Create and deploy a replacement before revoking the old key.",
  "rejected": "Revoke the old key before creating a replacement."
}
```

Omitting `objective` must continue to mean SFT.

## Motivation

The frozen Unsloth/TRL stack contains `DPOTrainer` and `DPOConfig`, but AIKit constructs only `SFTTrainer` and always converts data through the Alpaca path. DPO uses preferred/rejected responses rather than demonstrated completions and needs a separate trainer, objective-specific defaults, validation, and end-to-end export testing.

DPO should precede ORPO or KTO because it establishes the canonical preference-data and objective-dispatch foundation.

## Proposed changes

- Add optional top-level objective configuration; omitted objective defaults to `sft`.
- Add `objective.type: dpo` and `type: preference`.
- Initially require explicit, non-empty string `prompt`, `chosen`, and `rejected` values.
- Reject identical chosen/rejected responses and do not infer implicit prompts initially.
- Add and validate:
  - Finite positive `beta`.
  - `lossType: sigmoid` initially.
  - Positive `maxPromptLength` no larger than `maxSeqLength`.
  - DPO-specific learning-rate default rather than the current SFT default.
  - `packing: false` requirement.
- Import Unsloth before TRL and dispatch to `DPOTrainer`/`DPOConfig`.
- Use the PEFT policy model with `ref_model=None` so the disabled adapter supplies reference behavior.
- Never apply Alpaca, text, messages, or prompt-completion formatting to preference records.
- Load and validate dataset schema before GPU allocation where possible.
- Reuse the existing adapter/tokenizer save and GGUF export boundary.
- Include objective and DPO settings in the serialized training config/cache key.
- Reject unsupported format/objective combinations.

## Acceptance criteria

- Existing configurations without `objective` retain identical SFT behavior.
- DPO requires `type: preference`; preference data is rejected under SFT.
- Preference columns reach trainer construction unchanged.
- The Alpaca formatter is never called for DPO.
- Invalid/missing fields, identical choices, beta, length, packing, or loss settings fail clearly.
- DPO uses objective-specific defaults and `ref_model=None` with the PEFT policy model.
- Objective changes alter the training LLB digest.
- A one-step GPU DPO run saves and reloads the adapter/tokenizer, exports GGUF, and produces a non-empty inference response.
- Documentation clearly distinguishes SFT datasets from preference datasets.

## Tests

- Go objective defaulting and format/objective compatibility tests.
- Validation tests for beta, prompt length, loss type, learning-rate defaults, and packing.
- Python import-order and trainer-dispatch tests.
- Tests proving preference columns are preserved and the Alpaca formatter is bypassed.
- Schema tests for missing/null/empty/non-string fields and identical responses.
- Mock trainer assertions for `ref_model=None`, beta, loss type, and sequence lengths.
- Exact pinned-stack CPU preprocessing test.
- LLB serialization/digest tests.
- GPU DPO train/export/inference smoke test.

## Out of scope

- Conversational preference message arrays.
- Implicit prompt extraction.
- Multiple preference datasets.
- Additional or combined DPO loss types.
- User-selected reference models or reference-free DPO.
- Evaluation datasets/metrics.
- Full-parameter training.
- ORPO, KTO, PPO, GRPO, reward modeling, vision, or audio preference optimization.

### PR DESCRIPTION
**Pull request by sozercan**
Sibling stack on #833. This PR targets `issue-824`, is independent of #825, and should be merged after #833.

## Summary

- add an optional top-level fine-tuning objective with omitted/null/empty preserving existing SFT behavior
- add a separate DPO objective with pinned defaults for beta, sigmoid loss, prompt length, and learning rate
- add explicit `preference` records with validated prompt/chosen/rejected values
- enforce strict SFT/DPO dataset compatibility, one preference dataset, no packing, and no response-only SFT loss for DPO
- reuse #824 loader, immutable revision, checksum, cache-hardening, and URL-redaction behavior
- dispatch through the pinned Unsloth/TRL `DPOConfig` and `DPOTrainer` path with a PEFT policy and `ref_model=None`
- bypass all SFT formatting/masking and assert effective runtime DPO/reference-model contracts before training
- reuse the existing adapter/tokenizer save and GGUF export path
- serialize objective settings only at the DPO training boundary while preserving default-SFT LLB/cache identity
- add checksummed JSON DPO GPU smoke coverage, fixtures, tests, and documentation

## Tests

- `make test`
- `make test-python` (144 tests)
- focused Go objective/defaulting/validation/LLB tests
- exact pinned Unsloth 2026.8.3, TRL 0.24.0, and PEFT 0.20.0 contract validation
- exact-lock check with uv 0.12.1
- changed-lines `golangci-lint` (0 issues)
- actionlint/ShellCheck integration and workflow fixture validation
- `go mod tidy` clean
- independent DPO/SFT compatibility and cache-boundary review

The `dpo-smoke` GPU fine-tune workflow will be dispatched on this branch.

Closes #826

## GPU validation

- The first GPU run exposed Go YAML scientific notation (`1e-06`) being parsed by PyYAML as a string; commit `fa30a5c` adds strict numeric normalization and regression coverage.
- `dpo-smoke` then passed the preference load, one-step DPO train, adapter/tokenizer save, GGUF export, image build, and inference path on commit `fa30a5c`: [run 31176520125](https://github.com/kaito-project/aikit/actions/runs/31176520125).

### CODE REVIEW

File: pkg/finetune/target_unsloth.py [hunk 1]
```diff
@@ -3758,6 +4041,58 @@ def train_model(
...
+                report_to="none",
+            ),
+        )
+        validate_dpo_trainer_contract(
+            trainer,
+            policy_model=model,
+            objective=objective,
+            max_seq_length=max_seq_length,
+        )
+        trainer.train()
```
[the remark this answers is not in the document]
Comment (sozercan) [reply pkg/finetune/target_unsloth.py:4754]: Confirmed. With the pinned TRL 0.24.0 path, the trainer tokenizes each preference and later applies `keep_end` truncation to the combined prompt-plus-completion sequence. Distinct source responses can therefore become token-identical effective completions when their distinguishing prefix is truncated, eliminating the useful preference signal.

Commit `a1bc1dc` now validates the trainer-prepared token sequences, verifies the expected truncation contract, and rejects collapsed pairs before `train()`. It also adds a regression test and documents the behavior.
